aws-cpp-sdk-core-tests/utils/threading/ExecutorTest.cpp

```diff
@@ -0,0 +1,249 @@
+{
+    AWS_BEGIN_MEMORY_TEST(16, 10)
+    
+    numReceivedTasks = 0;
+    numFinishedTasks = 0;
+    const size_t POOL_SIZE = 3;
+    
+    std::shared_ptr<MockExecutor> mockExecutor =
+            Aws::MakeShared<MockExecutor>(EXECUTOR_TEST_ALLOCATION_TAG);
+    BlockingExecutor blockingExecutor(mockExecutor, POOL_SIZE);
+    
+    Aws::Vector<u_long> taskOrder;
+    std::mutex vectorLock;
+    std::condition_variable taskSignal;
+    
+    std::atomic<bool> lastTaskWaiting;
+    lastTaskWaiting = false;
+    
+    /* Each tasks waits for a "task signal" before proceeding. The last task will set a
+     * special flag. */
+    auto executeTask = [&](u_long id)
+    {
+        std::unique_lock<std::mutex> locker(vectorLock);
+        if (id == FINAL_SIZE - 1)
+        {
+            lastTaskWaiting = true;
+        }
+        taskSignal.wait(locker);
+        taskOrder.emplace_back(id);
+    };
+    
+    // Submit POOL_SIZE tasks to fill up the thread pool
+    blockingExecutor.Submit(executeTask, 0);
+    blockingExecutor.Submit(executeTask, 1);
+    blockingExecutor.Submit(executeTask, 2);
+    
+    // Submit an additional task. This one should get blocked, so we submit it in a separate thread.
+    auto submitLastTask = [&] ()
+    {
+        blockingExecutor.Submit(executeTask, 3);
+    };
+    std::thread t(submitLastTask);
+    t.detach();
+    
+    /* Check that we've received all tasks but the last one (since it should have been blocked
+     * in the blocking executor before being received by the mock executor) */
+//    ASSERT_TRUE(*mockExecutor->GetNumReceivedTasks() == CHECKPOINT_SIZE);
+    ASSERT_TRUE(numReceivedTasks == CHECKPOINT_SIZE);
+    std::unique_lock<std::mutex> locker(vectorLock);
+    ASSERT_TRUE(taskOrder.size() == 0);
+    locker.unlock();
+    
+    // Signal the currently queued tasks to run.
+    //    mockExecutor->GetTaskSignal()->notify_all();
+    taskSignal.notify_all();
+    locker.lock();
+    
+    checkpointSignal.wait(locker);
+    
+    // Check that all those tasks ran successfully.
+    ASSERT_TRUE(taskOrder.size() == CHECKPOINT_SIZE);
+    for (size_t i = 0; i < taskOrder.size(); i++)
+    {
+        ASSERT_FALSE(taskOrder.at(i) == 3);
+    }
+    locker.unlock();
+    
+    /* Additionally, the last task should now have been queued up (i.e. received by the mock executor
+     * , now that the others have finished. */
+//    ASSERT_TRUE(*mockExecutor->GetNumReceivedTasks() == FINAL_SIZE);
+    ASSERT_TRUE(numReceivedTasks == FINAL_SIZE);
+    
+    // Busy wait for the last task to let us know it's ready, then signal it.
+    while (!lastTaskWaiting);
+    //    mockExecutor->GetTaskSignal()->notify_all();
+    taskSignal.notify_all();
+    
+    locker.lock();
+    doneSignal.wait(locker);
+    
+    // Check completion and ordering of tasks.
+    ASSERT_TRUE(taskOrder.size() == FINAL_SIZE);
+    ASSERT_TRUE(taskOrder.at(3) == 3);
+    locker.unlock();
+    
+    AWS_END_MEMORY_TEST
+}
```

Comment:
It would seem good to do a test where you run (say) 3x pool size tasks and ensure they are actually running concurrently. I'd set pool size = 10 and have a task that sleeps 100ms. Then, you can make sure the 30 tasks take < 10 * 100ms (actually it should be about 300 ms but you want a little insurance).
I am not sure if this will be deemed too ugly for a test but at least good for a unit test!